staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,587 @@
+				propSchema: func(key string) *structuralschema.Structural { schema := schema.Properties[key]; return &schema },
+			}
+		}
+		if schema.AdditionalProperties != nil && schema.AdditionalProperties.Structural != nil {
+			return &unstructuredMap{
+				value:      m,
+				schema:     schema,
+				propSchema: func(key string) *structuralschema.Structural { return schema.AdditionalProperties.Structural },
+			}
+		}
+		return types.NewErr("invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema")
+	}
+	if schema.Type == "array" {
+		if schema.Items == nil {
+			return types.NewErr("invalid array type, expected Items with a non-empty Schema")
+		}
+		l, ok := unstructured.([]interface{})
+		if !ok {
+			return types.NewErr("invalid data, expected []interface{} to match the provided schema with type=array")
+		}
+		typedList := unstructuredList{elements: l, itemsSchema: schema.Items}
+		listType := schema.XListType
+		if listType != nil {
+			switch *listType {
+			case "map":
+				mapKeys := schema.Extensions.XListMapKeys
+				return &unstructuredMapList{unstructuredList: typedList, escapedKeyProps: escapeKeyProps(mapKeys)}
+			case "set":
+				return &unstructuredSetList{unstructuredList: typedList}
+			case "atomic":
+				return &typedList
+			default:
+				return types.NewErr("invalid x-kubernetes-list-type, expected 'map', 'set' or 'atomic' but got %s", *listType)
+			}
+		}
+		return &typedList
+	}
+
+	if schema.Type == "string" {
+		str, ok := unstructured.(string)
+		if !ok {
+			return types.NewErr("invalid data, expected string, got %T", unstructured)
+		}
+		if schema.ValueValidation != nil {
+			switch schema.ValueValidation.Format {
+			case "duration":
+				d, err := strfmt.ParseDuration(str)
+				if err != nil {
+					return types.NewErr("Invalid duration %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date":
+				d, err := time.Parse(strfmt.RFC3339FullDate, str) // strfmt uses this format for OpenAPIv3 value validation
+				if err != nil {
+					return types.NewErr("Invalid date formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(d)
+			case "date-time":
+				d, err := strfmt.ParseDateTime(str)
+				if err != nil {
+					return types.NewErr("Invalid date-time formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue(time.Time(d))
+			case "byte":
+				base64 := strfmt.Base64{}
+				err := base64.UnmarshalText([]byte(str))
+				if err != nil {
+					return types.NewErr("Invalid byte formatted string %s: %v", str, err)
+				}
+				return types.DefaultTypeAdapter.NativeToValue([]byte(base64))
+			}
+		}
+		return types.DefaultTypeAdapter.NativeToValue(unstructured)
+	}
+	if schema.Type == "number" || schema.Type == "integer" || schema.Type == "boolean" {
+		return types.DefaultTypeAdapter.NativeToValue(unstructured)
+	}
+	if schema.XPreserveUnknownFields {
```

Comment:
I thought XIntOrString was the only case that was allowed to omit type... XPreserveUnknownFields shouldn't allow type to be unspecified

Reply:
nope, I was wrong...

```
} else if len(s.Type) == 0 && !s.Extensions.XIntOrString && !s.Extensions.XPreserveUnknownFields {
```

Reply:
so... if type is unspecified... this could be a map, an array, or a scalar? I'm trying to wrap my head around what the right data to return here is... what does returning empty map data do?

Reply:
This is the equality issue I mentioned in the above comment. Right now equals would be false for `self[0] == self[1]` and true for `self[0].name == self[1].name`.  The gotcha is that if we decide that we can't reason about unknown values, then even if both have `foo: true`, I don't think we should conclude they are equal?  E.g. foo is a list of objects, do we know if it's an associative list or a normal list? How do we know if it's equal or not.  For this particular case, it's a bit harder to say with a straight face that we don't know if they're equal or not, but I don't know exactly where to draw the line on what we include when do a compare.